Lib/test/test_tarfile.py

```diff
@@ -1878,6 +1878,40 @@ def test_longnamelink_1025(self):
         self._test(("longnam/" * 127) + "longname_",
                    ("longlnk/" * 127) + "longlink_")
 
+    def test_hidden_header_for_longname(self):
+        # Regression test for gh-130819.
+        memory_file = io.BytesIO()
+        tar = tarfile.open(mode="w", fileobj=memory_file, format=tarfile.GNU_FORMAT)
+        tar_info = tarfile.TarInfo("abcdef" * 20)
+        tar_info.type = tarfile.DIRTYPE
+        tar.addfile(tar_info, None)
+        tar.close()
+
+        class RawTabInfo(tarfile.TarInfo):
+
+            def _proc_member(self, tar_file):
+                if self.type in (tarfile.GNUTYPE_LONGNAME, tarfile.GNUTYPE_LONGLINK):
+                    tester.assertEqual(self.mode, 0o644)
+                    unames_cache = RawTabInfo._unames
+                    gnames_cache = RawTabInfo._gnames
+                    if unames_cache:
+                        tester.assertIn(0, unames_cache)
+                    if gnames_cache:
+                        tester.assertIn(0, gnames_cache)
+                    tester.assertEqual(self.uname, unames_cache.get(0, ""))
+                    tester.assertEqual(self.gname, gnames_cache.get(0, ""))
+                return super()._proc_member(tar_file)  # type: ignore
+
+        tester = self
+        memory_file.seek(0)
+        tar = tarfile.open(fileobj=memory_file, mode="r", tarinfo=RawTabInfo)
+        try:
+            members = tar.getmembers()
+            self.assertEqual(len(members), 1)
+        finally:
+            tar.close()
+            memory_file.close()
```

Comment:
```suggestion
        tar = tarfile.open(fileobj=memory_file, mode="r", tarinfo=RawTabInfo)
        self.addCleanup(tar.close)
        members = tar.getmembers()
        self.assertEqual(len(members), 1)
```

The memory file does not need to be closed as it's in-memory only.